QAI-191, 192, 193, 194, 195, 197: ETL Pipeline for BBC News

karthikstar opened the pull request:
For the following user stories:
QAI-191, QAI-192, QAI-193, QAI-194, QAI-195, QAI-197
- Collection with minimum viable schema implemented
- Successful retrieval of all data rows from BBC News API for all months from 2017-01 to 2024-09 (total of ~125k rows)
- Successful insertion of all raw data into a MongoDB collection
- Transformation pipeline which filters down to relevant columns only and retain rows with non-missing values for title, publishedDate, description, content
- Successful loading of transformed rows back into another MongoDB Collection
- Implemented a GET endpoint on middleware which returns users all the transformed BBC articles. This is facilitated by communication between ETL and Middleware via REST.

For reference:
Pre-transformation rows loaded in: 124750
Post-Transformation rows loaded into DB: 124731

karthikstar commented:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news/fetch  

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/transform_bbc/run

karthikstar commented:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/transform_bbc/run

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/api/transformed-bbc-news/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/bbcnews/transformed

Ivsucram commented on build.gradle:
```diff
@@ -38,7 +38,12 @@ repositories {
     implementation("org.springframework.boot:spring-boot-starter-web")
+    implementation("org.springframework.boot:spring-boot-starter-security")
+    implementation("org.springframework.boot:spring-boot-configuration-processor")
+    implementation("org.springframework.boot:spring-boot-starter-web")
```
"implementation("org.springframework.boot:spring-boot-starter-web")" is being imported twice.

karthikstar replied:
fixed

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt:
```diff
@@ -0,0 +1,23 @@
+package sg.com.quantai.etl.controllers
```
Add a space (empty line) between the package and imports, making it easy to visualize quickly.

Ivsucram replied:
Add a space (empty line) between the package and imports, making it easy to visualize quickly.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt:
```diff
@@ -0,0 +1,53 @@
+package sg.com.quantai.etl.controllers
```
Add a space (empty line) between the package and imports, making it easy to visualize quickly.

Ivsucram replied:
Add a space (empty line) between the package and imports, making it easy to visualize quickly.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/controllers/BbcNewsTransformationController.kt:
```diff
@@ -0,0 +1,23 @@
+
+@RestController
+@RequestMapping("/transform_bbc")
+class BbcNewsTransformationController(
```
Lines 9 and 10

The scope of the data is still the same: BBC News.
Therefore, a new controller is not necessary.

Ivsucram replied:
Lines 9 and 10

The scope of the data is still the same: News Headlines.
Therefore, a new controller is not necessary.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/controllers/BbcRawNewsController.kt:
```diff
@@ -0,0 +1,53 @@
+
+@RestController
+@RequestMapping("/news")
+class BbcRawNewsController(
```
Line 12:
This endpoint does not fully reflect the data flow. Use something like "news_headlines," and then organize the data flow via functions and endpoints.

Line 13:
The scope here is still News Headlines. There shouldn't be a Raw Controller, but only a News Headline Controller with specific endpoints and functions to handle the different data formats (raw, transformed) and origin (we have a single source at this moment: BBC).

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/controllers/BbcTransformedNewsController.kt:
```diff
@@ -0,0 +1,24 @@
+
+@RestController
+@RequestMapping("/api/transformed-bbc-news")
+class BbcTransformedNewsController(
```
Line 10:
This endpoint does not fully reflect the data flow. Use something like "news_headlines," then organize the data flow via functions and endpoints.

Line 11:
The scope here is still News Headlines. There shouldn't be a Transformed or Specific API Controller, but only a News Headline Controller with specific endpoints and functions to handle the different data formats (raw, transformed), origin (we have a single source at this moment: BBC), and behaviors (fetch from external, deliver to internal call, etc).

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt:
```diff
@@ -0,0 +1,17 @@
+import java.time.LocalDate
+
+@Document(collection = "bbc_news_articles_transformed")
+data class BbcTransformedNewsArticle(
```
The transformed data class should be an abstraction in the format necessary for the platform's operations. The platform (users, researchers, and internal calls) does not currently care about the origin of the transformed data. Therefore, this data class should be something like "news_headlines."

I know that we only have one data source: BBC News. But, as discussed before, if we have multiple data sources, we can use them to auto-complete each other's information and create a single transformed data record that fulfills the needs of our platform.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.data.mongodb.core.mapping.Document
+import java.time.LocalDate
+
+@Document(collection = "bbc_news_articles_raw")
```
If you call the other data collection "news_headlines," call this something like "news_headlines_BBC" to indicate that it is a news_headlines of type BBC (where BBC automatically denotes the origin of the raw data).

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:
```diff
@@ -0,0 +1,25 @@
+import java.time.LocalDate
+
+@Document(collection = "bbc_news_articles_raw")
+data class BbcRawNewsArticle(
```
If you call the other data class "NewsHeadlines," call this something like "NewsHeadlinesBBC" to indicate that this is NewsHeadliens of type BBC (where BBC automatically denotes the origin of the raw data).

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcTransformedNewsArticle.kt:
```diff
@@ -0,0 +1,17 @@
+
+@Document(collection = "bbc_news_articles_transformed")
+data class BbcTransformedNewsArticle(
+    val title: String,
```
As suggested via written message on Discord, please use the format specified at https://github.com/Quant-NTU/persistence-middleware/blob/main/src/main/kotlin/sg/com/quantai/middleware/data/NewStrategy.kt to organize your data class

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:
```diff
@@ -0,0 +1,25 @@
+@Document(collection = "bbc_news_articles_raw")
+data class BbcRawNewsArticle(
+    val title: String,
+    val publishedDate: LocalDate,
```
As suggested via written message on Discord, please use the format specified at https://github.com/Quant-NTU/persistence-middleware/blob/main/src/main/kotlin/sg/com/quantai/middleware/data/NewStrategy.kt to organize your data class

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/repositories/BbcTransformedNewsRepository.kt:
```diff
@@ -0,0 +1,7 @@
+package sg.com.quantai.etl.repositories
+
+import sg.com.quantai.etl.data.BbcTransformedNewsArticle
```
This class does not exist

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/requests/BbcRawNewsArticleRequest.kt:
```diff
@@ -0,0 +1,12 @@
+package sg.com.quantai.etl.requests
+
+class BbcRawNewsArticleRequest(
```
Change the name of this class accordingly to the name of the refactored data class..

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+    @Autowired private val bbcRawNewsRepository: BbcRawNewsRepository
+) {
+
+    private val newApiUrl = "https://datasets-server.huggingface.co/"
```
The url string should be in the properties file.
The variable name "newApiUrl" should be updated accordingly.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+import java.time.Duration
+
+@Service
+class BbcNewsApiService(
```
The class name should be updated accordingly, such as "NewsHeadlineBBCService"

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+    private val newApiUrl = "https://datasets-server.huggingface.co/"
+    private val batchSize = 100
+
+    private fun bbcNewsWebClient(): WebClient {
```
the method name can be changed to just webClient without losing its meaning

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+            .build()
+    }
+
+    fun fetchAndSaveNewsArticlesForAllMonths() {
```
Refactor title to "fetchAndSaveFromAllMonths"

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+        println("Loading of BBC News Articles - START")
+
+        val startDate = YearMonth.of(2017, 1)
+        val endDate = YearMonth.of(2024, 9)
```
End Date should be set programmatically, and not static

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+        var currentMonth = startDate
+        while (currentMonth <= endDate) {
+            val configDate = currentMonth.toString()
+            println("Fetching data for: $configDate")
```
Avoid using print, println and others. Use logs instead.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+            fetchAndSaveNewsArticlesForDate(configDate)
+            currentMonth = currentMonth.plusMonths(1)
+        }
+        println("Loading of BBC News Articles - END")
```
Avoid using print, println and others. Use logs instead.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+        println("Loading of BBC News Articles - END")
+    }
+
+    private fun fetchAndSaveNewsArticlesForDate(config: String) {
```
Refactor title to "fetchAndSaveFromDate."
Use a more descriptive data type for config, such as java.time
User a more descriptive name for the variable as well, as config is too generic and doesn't give any information that you are retrieving data related to a date range

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+            }
+
+            println("Successfully saved all articles for $config.")
+        } catch (ex: Exception) {
```
Create a custom exception

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+                        }
+                )
+                .block() ?: throw RuntimeException("Failed to fetch rows")
+        } catch (ex: Exception) {
```
Create a custom exception

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    private fun fetchRows(offset: Int, length: Int, config: String): String {
```
Use a more descriptive data type for config, such as java.time
User a more descriptive name for the variable as well, as config is too generic and doesn't give any information that you are retrieving data related to a date range

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+                                    throwable.statusCode.is5xxServerError
+                        }
+                )
+                .block() ?: throw RuntimeException("Failed to fetch rows")
```
throw custom exception

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsApiService.kt:
```diff
@@ -0,0 +1,118 @@
+        return objectMapper.readTree(response)
+    }
+
+    private fun processRows(rows: JsonNode) {
```
Refactor the title to something more descriptive.
What does process mean?

This method is inserting records in the database, so use something related to this action.

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/services/BbcNewsTransformationService.kt:
```diff
@@ -0,0 +1,38 @@
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import sg.com.quantai.etl.data.BbcRawNewsArticle
+import sg.com.quantai.etl.data.BbcTransformedNewsArticle
```
Imported class does not exist

karthikstar commented:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/transform   

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/news_headlines/bbc/api/news_transformed

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_headlines/bbc_transformed/all

karthikstar commented:
- Requested changes have been made
- Added unit tests for NewsHeadlinesController

Updated endpoints to check during functional test:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_headlines/bbc/transform

To retrieve all transformed BBC news from database (within ETL):
curl -X GET http://localhost:10070/news_headlines/bbc/api/news_transformed

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_headlines/bbc_transformed/all

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/data/BbcRawNewsArticle.kt:
Line 9: 
The collection is still following the wrong name convention. It should be "news_articles_bbc," where "news_articles" denotes the scope that we intend to store, and "BBC" already denotes that it stores BBC raw data. This convention should also be described in the documentation.

Ivsucram replied:
Line 10:
The data class name still follows the wrong name convention. It should be "NewsArticleBBC."

Ivsucram commented on src/main/kotlin/sg/com/quantai/etl/repositories/BbcRawNewsRepository.kt:
The data class name still follows the wrong name convention. It should be "NewsArticleBBC."

Ivsucram commented on src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    fun `fetchAndSaveNews should return OK status`() {
```
The title of the test should be fully English formatted. So "should connect to fetch results" should suffice.

Ivsucram commented on src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    fun `runTransformation should return OK when successful`() {
```
The title of the test should be fully English formatted to describe which action and reaction is expected. So something like "Should apply transformations on BBC data" should suffice.

Ivsucram commented on src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    fun `runTransformation should return INTERNAL_SERVER_ERROR when transformation fails`() {
```
The test title should be fully English formatted to describe which action and reaction is expected. So something like "Should yield proper exception if a transformation fails" or similar would be better.

Ivsucram commented on src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    fun `getAllTransformedNews should return articles when available`() {
```
The test title should be fully English formatted to describe which action and reaction is expected. I am not sure of the goal of this test by the title.

Ivsucram commented on src/test/kotlin/sg/com/quantai/etl/controllers/NewsHeadlinesControllerTest.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Test
+    fun `getAllTransformedNews should return NO_CONTENT when no articles are available`() {
```
The test title should be fully English formatted to describe which action and reaction is expected. So something like "A NO_CONTENT response should be given if there are no content available" or something.

karthikstar commented:
Updated: 
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all

karthikstar commented:
- Requested code changes done.
- Fixed Unit Tests. All passed on my end

You can test the functionality using these updated endpoints:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all

karthikstar commented:
Made code changes to:
- Skip fields 'author' and 'section'
- Check before inserting the data, in order to handle for multiple runs of fetch data

You can test the functionality using these updated endpoints:
To run initial fetching of raw data and loading into DB:
curl -X POST http://localhost:10070/news_articles/bbc/fetch

To run transformation pipeline and loading transf. rows into DB:
curl -X POST http://localhost:10070/news_articles/bbc/transform

To retrieve all transformed news from database (within ETL):
curl -X GET http://localhost:10070/news_articles/api/all

To retrieve all transformed BBC news from database (within middleware):
curl -X GET http://localhost:10001/news_articles/all
